x-pack/plugins/security_solution/server/lib/entity_analytics/asset_criticality/asset_criticality_service.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { isEmpty } from 'lodash/fp';
+import type { ExperimentalFeatures } from '../../../../common';
+import type { AssetCriticalityRecord } from '../../../../common/api/entity_analytics';
+import type { AssetCriticalityDataClient } from './asset_criticality_data_client';
+
+interface CriticalityIdentifier {
+  id_field: string;
+  id_value: string;
+}
+
+interface IdentifierValuesByField {
+  [idField: string]: string[];
+}
+
+export interface AssetCriticalityService {
+  getCriticalitiesByIdentifiers: (
+    identifiers: CriticalityIdentifier[]
+  ) => Promise<AssetCriticalityRecord[]>;
+  isEnabled: () => boolean;
+}
+
+const isCriticalityIdentifierValid = (identifier: CriticalityIdentifier): boolean =>
+  !isEmpty(identifier.id_field) && !isEmpty(identifier.id_value);
+
+const groupIdentifierValuesByField = (
+  identifiers: CriticalityIdentifier[]
+): IdentifierValuesByField =>
+  identifiers.reduce((acc, id) => {
+    if (!acc[id.id_field]) {
+      acc[id.id_field] = [];
+    }
+    if (!acc[id.id_field].includes(id.id_value)) {
+      acc[id.id_field].push(id.id_value);
+    }
+    return acc;
+  }, {} as IdentifierValuesByField);
+
+const buildCriticalitiesQuery = (identifierValuesByField: IdentifierValuesByField) => ({
+  bool: {
+    filter: {
+      bool: {
+        should: Object.keys(identifierValuesByField).map((idField) => ({
+          bool: {
+            must: [
+              { term: { id_field: idField } },
+              { terms: { id_value: identifierValuesByField[idField] } },
+            ],
+          },
+        })),
+      },
+    },
+  },
+});
+
+const getCriticalitiesByIdentifiers = async ({
+  assetCriticalityDataClient,
+  identifiers,
+}: {
+  assetCriticalityDataClient: AssetCriticalityDataClient;
+  identifiers: CriticalityIdentifier[];
+}): Promise<AssetCriticalityRecord[]> => {
+  if (identifiers.length === 0) {
+    throw new Error('At least one identifier must be provided');
+  }
+  const validIdentifiers = identifiers.filter((id) => isCriticalityIdentifierValid(id));
+
+  if (validIdentifiers.length === 0) {
+    throw new Error('At least one identifier must contain a valid field and value');
```

Comment:
Why did you decide to return an error here, and not just an empty array?

Reply:
I think it was partly my expectation that this would correspond to an API route, and the lack of confidence in the `identifiers` type. I wanted useful feedback for the user if they provided bad data into this, and with an empty array it wouldn't be clear if that was due to no criticality records being found, or bad input data. 

The `search` below also has the potential to throw an error, so it didn't seem like a big leap to add another error here. I'm happy to change it around, but maybe JSDoc mentioning the error-throwing behavior would be helpful?